src/modalities/models/coca/coca_model.py

```diff
@@ -0,0 +1,142 @@
+            sample_key=text_embd_prediciton_key,
+            prediction_key=text_decoder_config.prediction_key,
+            block_size=text_decoder_config.block_size,
+            n_layer=n_shared_layer,
+            **shared_decoder_kwargs,
+        )
+
+        # TODO Validate if weight tying is useful for coca
+        self.text_decoder.transformer.wte.weight = (
+            self.multimodal_decoder.lm_head.weight
+        )  # https://paperswithcode.com/method/weight-tying
+
+        # vision_queries: 256 queries for multimodal cross attention and 1 as vision cls token for contrastive learning
+        self.vision_queries = nn.Parameter(torch.randn(n_vision_queries + 1, vision_encoder_config.n_embd))
+        self.attn_pool = AttentionPooling(
+            n_embd=vision_encoder_config.n_embd,
+            n_head=n_pool_head,
+            bias=bias_attn_pool,
+            epsilon=epsilon_attn_pool,
+        )
+
+        # init all weights
+        weight_init = text_decoder_config.weight_init
+        self.apply(partial(self._init_weights, weight_init=weight_init))
+        # apply special scaled init to the residual projections, per GPT-2 paper
+        for pn, p in self.named_parameters():
+            if pn.endswith("c_proj.weight"):
+                torch.nn.init.normal_(
+                    p, mean=weight_init.mean, std=weight_init.std / math.sqrt(2 * text_decoder_config.n_layer)
+                )
+
+    def _init_weights(self, module: nn.Module, weight_init: WeightInitailizationConfig):
+        if isinstance(module, nn.Linear):
+            torch.nn.init.normal_(module.weight, mean=weight_init.mean, std=weight_init.std)
+            if module.bias is not None:
+                torch.nn.init.zeros_(module.bias)
+        elif isinstance(module, nn.Embedding):
+            torch.nn.init.normal_(module.weight, mean=weight_init.mean, std=weight_init.std)
+
+    def forward(self, inputs):
+        vision_embd, vision_cls_token = self._forward_encode_vision(inputs)
+        text_embd, text_cls_token = self._forward_encode_text(inputs)
+        logits = self._forward_decode(text_embd, vision_embd)
+        return {
+            self.prediction_key: logits,
+            self.vision_cls_prediciton_key: vision_cls_token,
+            self.text_cls_prediciton_key: text_cls_token,
+        }
+
+    def _forward_encode_vision(self, inputs):
+        vision_embd = self.vision_encoder(inputs)[self.vision_embd_prediciton_key]
+        queries = repeat(self.vision_queries, "n d -> b n d", b=vision_embd.shape[0])
+        vision_embd = self.attn_pool(vision_embd, queries)
+        vision_cls_token, vision_embd = vision_embd[:, :1, :], vision_embd[:, 1:, :]
+        return vision_embd, vision_cls_token
+
+    def _forward_encode_text(self, inputs):
+        text_embd = self.text_decoder(inputs)[self.text_embd_prediciton_key]
+        text_embd, text_cls_token = text_embd[:, :-1, :], text_embd[:, -1:, :]

```

Comment:
This has no side effects because we first "build" the tuple, store it, and then override the variables. Side effects could only occure if we would use `text_embd` to index the output variables. See https://stackoverflow.com/questions/34171348/tuple-unpacking-order-changes-values-assigned